### ISSUE TITLE
Linux: Some window managers automatically move windows between monitors if the window doesn't fit

### ISSUE THREAD

**Issue by gdrewb-valve** [ValveSoftware/Dota-2#556]
When creating a window that has the same size as a monitor the window may not fit due to window decorations or space taken for a dock/tray/etc.  Some window managers will move the window to a monitor where it can fit in that case.  Reborn will not have the correct monitor info as it doesn't realize that it did not get its window in the place it wanted.

Splitting this specific case from issue #13, but all of the details are there.

**Comment by serianx** [comment]
This is affecting me, I've read the comments on issue #13 and I want to help fix this bug. 

This is all the information I have gathered:
I use gnome3
Usually, my primary monitor is on the right  (check SDL test at the end)
If I force dota to open in windowed mode, then it will open on the correct monitor, ( it also changes, according to the -sdl_displayindex value)
if the window is set to borderless < 1920x1080 then it also works fine.
if it is set on borderless 1080p or desktop friendly fullscreen, it will open on the secondary monitor,
no matter what -sdl_displayindex value I set
HOWEVER, if i don't set any value, or I set it =0  the window will open in the secondary monitor AND
become unresponsive, the game freezes and i have to kill the process. This is maybe related with the
freeze bugs also reported on github?

If I change the default monitor to be the left one in gnome display settings, 
the behaviour is similar (fullscreen or borderless 1080p showing on secondary monitor, right monitor this time)  but the freezes are not there! the game works fine on the secondary monitor.

-sdl_displayindex 2 doesn't work for me, it doesn't make the game go to the correct monitor, I've tried 3 aswell. all it does is make the game appear on the corner instead of centered, but on the wrong monitor.

the output of the SDL test is:

```
INFO: Using video target 'x11'.
INFO: See 2 displays.
INFO: 0: "22MP55 22"" (1920x1080, (1920, 0)), 17 modes.
INFO: CURRENT: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
INFO: DESKTOP: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
INFO:     MODE 0: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
INFO:     MODE 1: fmt=SDL_PIXELFORMAT_RGB888 w=1680 h=1050 refresh=59
INFO:     MODE 2: fmt=SDL_PIXELFORMAT_RGB888 w=1600 h=900 refresh=60
INFO:     MODE 3: fmt=SDL_PIXELFORMAT_RGB888 w=1440 h=900 refresh=59
INFO:     MODE 4: fmt=SDL_PIXELFORMAT_RGB888 w=1400 h=1050 refresh=59
INFO:     MODE 5: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=1024 refresh=75
INFO:     MODE 6: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=1024 refresh=60
INFO:     MODE 7: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=800 refresh=59
INFO:     MODE 8: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=720 refresh=60
INFO:     MODE 9: fmt=SDL_PIXELFORMAT_RGB888 w=1152 h=864 refresh=75
INFO:     MODE 10: fmt=SDL_PIXELFORMAT_RGB888 w=1024 h=768 refresh=75
INFO:     MODE 11: fmt=SDL_PIXELFORMAT_RGB888 w=1024 h=768 refresh=60
INFO:     MODE 12: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=75
INFO:     MODE 13: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=60
INFO:     MODE 14: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=56
INFO:     MODE 15: fmt=SDL_PIXELFORMAT_RGB888 w=640 h=480 refresh=75
INFO:     MODE 16: fmt=SDL_PIXELFORMAT_RGB888 w=640 h=480 refresh=59
INFO: 
INFO: 1: "DELL S2240L 21"" (1920x1080, (0, 0)), 22 modes.
INFO: CURRENT: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
INFO: DESKTOP: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
INFO:     MODE 0: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
INFO:     MODE 1: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=59
INFO:     MODE 2: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=50
INFO:     MODE 3: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=30
INFO:     MODE 4: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=29
INFO:     MODE 5: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=25
INFO:     MODE 6: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=1024 refresh=75
INFO:     MODE 7: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=1024 refresh=60
INFO:     MODE 8: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=720 refresh=60
INFO:     MODE 9: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=720 refresh=59
INFO:     MODE 10: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=720 refresh=50
INFO:     MODE 11: fmt=SDL_PIXELFORMAT_RGB888 w=1152 h=864 refresh=75
INFO:     MODE 12: fmt=SDL_PIXELFORMAT_RGB888 w=1024 h=768 refresh=75
INFO:     MODE 13: fmt=SDL_PIXELFORMAT_RGB888 w=1024 h=768 refresh=60
INFO:     MODE 14: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=75
INFO:     MODE 15: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=60
INFO:     MODE 16: fmt=SDL_PIXELFORMAT_RGB888 w=720 h=576 refresh=50
INFO:     MODE 17: fmt=SDL_PIXELFORMAT_RGB888 w=720 h=576 refresh=25
INFO:     MODE 18: fmt=SDL_PIXELFORMAT_RGB888 w=720 h=480 refresh=59
INFO:     MODE 19: fmt=SDL_PIXELFORMAT_RGB888 w=720 h=480 refresh=30
INFO:     MODE 20: fmt=SDL_PIXELFORMAT_RGB888 w=640 h=480 refresh=75
INFO:     MODE 21: fmt=SDL_PIXELFORMAT_RGB888 w=640 h=480 refresh=59
INFO: 
```

What else can I do to help?

**Comment by gdrewb-valve** [comment]
You could try and contact the maintainers of your window manager and see what they say or if they have more info.  However at this point I think it'll have to wait until the SDL team can try and dig into it.

**Comment by Jusix** [comment]
Has someone already opened a bug @ the SDL bug tracker?

**Comment by gdrewb-valve** [comment]
I contacted the SDL devs directly.  There are already multiple fullscreen bugs open in the SDL bugzilla too.

**Comment by ruscur** [comment]
The workaround from #13, -sdl_displayindex 2, has regressed in the current test client.  Now always displays on the wrong monitor regardless of argument to sdl_displayindex.

**Comment by gdrewb-valve** [comment]
That's probably an SDL fix since using an invalid display index was never something that was supposed to help.  I'll pass this on to the SDL folks.

**Comment by gdrewb-valve** [comment]
@ruscur , what window manager are you using?

**Comment by ruscur** [comment]
cinnamon 2.8.3

**Comment by gdrewb-valve** [comment]
We've added an experiment in the March 24 update to try and get more data.  For people with problems can you add -phased_window_create to your launch options and see if that changes anything?

**Comment by ruliane** [comment]
Hi,
I just switch from Linux Mint 17.2 to 17.3, and things are getting even worse :
- In 17.2, when I switched to windowed mode, then back to fullscreen mode, I could play on my primary (right) screen.
- Now, the game is moved on the left screen if I select fullscreen mode.

Neither -phased_window_create nor -sdl_displayindex helped me.

**Comment by gdrewb-valve** [comment]
What is your window manager?

**Comment by ruliane** [comment]
Can't believe I forgot to mention... I'm using MATE 1.12.

**Comment by gdrewb-valve** [comment]
The March 31 update has a tweak to -phased_window_create, does it make any difference?  Also in the console output on initial window creation there should now be a line like

```
Create window 'Dota 2' at xxx,xxx - xxx,xxx, flags 0xxxx
```

Does that show up?  The four numbers are x,y - width,height.  Does the x,y coordinate look correct for the upper-left corner of the screen you want to use?  Does it change when you use -sdl_displayindex 0 vs. 1?

**Comment by msjyoo** [comment]
Hello, using Cinnamon on Debian 8 here. I have the same issue. Before if I went into windowed mode and switch back to desktop fullscreen, I could play on my primary monitor (left). But now, even that workaround doesn't seem to work and I'm now forced to play in windowed mode just to get the correct monitor :(

1920x1200
Dual monitor, left primary
Debian 8, Cinnamon DE

@gdrewb-valve

**Comment by Tele42** [comment]
@sekjun9878 cinnamon has a hotkey to move fullscreen windows from one monitor to another, which will work when the steam overlay is up.

**Comment by gdrewb-valve** [comment]
Does Dota initially start on the correct monitor?  Does -phased_window_create make a difference? Can you post the startup information I requested?

**Comment by dhiru1602** [comment]
I was suffering from Dota 2 launching on wrong monitor ever since the release of Reborn client. I was using the _-sdl_displayindex "some-invalid-value"_ to force it to launch Dota 2 on the primary monitor. This workaround no longer works with recent SDL update.

I tried -phased_window_create and removed the sdl_displayindex parameter and Dota 2 correctly launches on my primary monitor. I am also able to change the resolution settings and I see a list of all the supported resolutions (earlier it was "No supported resolutions"). I can confirm that the phased window creation workaround actually works. Atleast for me.

`Create window 'Dota 2' at 0,0 - 1920,1080, flags 0x1017`

Here is the output for SDL test.

> INFO: Using video target 'x11'.
> INFO: See 2 displays.
> INFO: 0: "IPS236        23"" (1920x1080, (0, 0)), 8 modes.
> INFO:     DPI: ddpi=95.372787; hdpi=95.623528; vdpi=94.593102
> INFO: CURRENT: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
> INFO: DESKTOP: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
> INFO:     MODE 0: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
> INFO:     MODE 1: fmt=SDL_PIXELFORMAT_RGB888 w=1680 h=1050 refresh=59
> INFO:     MODE 2: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=1024 refresh=60
> INFO:     MODE 3: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=960 refresh=60
> INFO:     MODE 4: fmt=SDL_PIXELFORMAT_RGB888 w=1152 h=864 refresh=59
> INFO:     MODE 5: fmt=SDL_PIXELFORMAT_RGB888 w=1024 h=768 refresh=60
> INFO:     MODE 6: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=60
> INFO:     MODE 7: fmt=SDL_PIXELFORMAT_RGB888 w=640 h=480 refresh=59
> INFO: 
> INFO: 1: "IPS236        23"" (1920x1080, (1920, 0)), 8 modes.
> INFO:     DPI: ddpi=95.372787; hdpi=95.623528; vdpi=94.593102
> INFO: CURRENT: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
> INFO: DESKTOP: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
> INFO:     MODE 0: fmt=SDL_PIXELFORMAT_RGB888 w=1920 h=1080 refresh=60
> INFO:     MODE 1: fmt=SDL_PIXELFORMAT_RGB888 w=1680 h=1050 refresh=59
> INFO:     MODE 2: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=1024 refresh=60
> INFO:     MODE 3: fmt=SDL_PIXELFORMAT_RGB888 w=1280 h=960 refresh=60
> INFO:     MODE 4: fmt=SDL_PIXELFORMAT_RGB888 w=1152 h=864 refresh=59
> INFO:     MODE 5: fmt=SDL_PIXELFORMAT_RGB888 w=1024 h=768 refresh=60
> INFO:     MODE 6: fmt=SDL_PIXELFORMAT_RGB888 w=800 h=600 refresh=60
> INFO:     MODE 7: fmt=SDL_PIXELFORMAT_RGB888 w=640 h=480 refresh=59
> INFO: 

**OS:** Arch Linux
**DE:** Gnome 2.8
**WM:** Mutter 3.18

So basically to summarize, I have set the phased_window_create launch option and Dota 2 always opens up on the primary display initially. When I select fullscreen, it switches to fullscreen on the current display that the game is currently running on. Everything works well as expected and resolutions always show up correctly. I have two identical monitors and the resolution support on both monitors is the same. I am not sure how it would work on multiple monitor setups with different resolutions on each display.

Thanks @gdrewb-valve!

**Comment by ruscur** [comment]
Finally reunited with my desktop so I can test this.

`-phased_window_create` works for me, no other arguments.

**Comment by msjyoo** [comment]
`-phased_window_create` launches the game on the primary window correctly for me. However, switching from Windowed to "Desktop fullscreen" in-game seems to make the game go back to the wrong behaviour again. `-sdl_displayindex 0` made no difference.

Still a huge improvement though, thanks for the fix!

**Comment by gdrewb-valve** [comment]
OK, sounds like -phased_window_create is working for everyone.  We'll hook some more of its logic in the video settings path and that will hopefully fix the case of switching from windowed to fullscreen.

**Comment by gdrewb-valve** [comment]
I think the windowed->fullscreen fixes have shipped, are you still seeing problems with that transition?

**Comment by dhiru1602** [comment]
Could you please confirm if **-phased_window_create** is still required? I am currently using BuildID:**1121941** and removing the **-phased_window_create** launch option causes Dota2 to launch on the secondary monitor. Adding it back again launches Dota 2 on the primary monitor.

**Comment by gdrewb-valve** [comment]
Yes, it's still required and likely will stay that way.

**Comment by gdrewb-valve** [comment]
No problem reports have come in so closing for now.

**Comment by ChrisLauinger77** [comment]
I needed to change also video settings ingame to "Desktop friendly full screen".
Then -phased_window_create worked 4 me, 2 :+1:

**Comment by ase1590** [comment]
with dota 2 7.00, it seems phased_window_create is no longer working.

**Comment by materemias** [comment]
unfortunately -phased_window_create conflicts with Vulkan

**Comment by LiamDawe** [comment]
Just to comment, that it took me forever to find this issue. Constantly starts on the non-primary monitor. It's still a problem on Ubuntu 17.10 with NVIDIA drivers.

-phased_window_create only works if it's set to "Desktop-friendly fullscreen" in the "Display Mode" dropdown.

Is it not possible to have this behaviour added to the options somehow? Having to dig through issues to find this was a real hassle.

**Comment by Apple-Lion** [comment]
It's back from the patch a couple weeks back.

**Comment by rcketscientist** [comment]
sdl and phased doesn't work for me on a 580 in Ubuntu 18.10.  This is honestly ridiculous.  How is there not at least a config to set to say "f it...hard code it here".

**Comment by alasher** [comment]
I still have this issue using Pop!_OS 19.04 on Nvidia with Gnome. But it does work with @LiamDawe's above suggestion of using -phased_window_create with "Desktop-friendly fullscreen".
